external/error_report/error_report.c

```diff
@@ -43,40 +46,147 @@ typedef struct {
 	uint16_t front;
 	int8_t q_pending;
 	pthread_mutex_t err_mutex;
+#if CONFIG_ERROR_REPORT_INFINITE_WAIT
+	sem_t infinity_loop_exit;
+	pthread_t infinity_thread;
+	uint8_t ntasks;
+#endif
 	err_fsm_t fsm;
 } g_error_info_t;
 
 static g_error_info_t g_err_info;
+static int g_err_report_fd;
 
-static unsigned long prv_fetch_taskaddr(int pid)
-{
-	struct tcb_s *tcbptr = sched_gettcb(pid);
-	if (tcbptr != NULL) {
-		entry_t e = tcbptr->entry;
-		if ((tcbptr->flags & TCB_FLAG_TTYPE_MASK) == TCB_FLAG_TTYPE_PTHREAD) {
-			return (unsigned long)e.pthread;
-		} else {
-			return (unsigned long)e.main;
+#if CONFIG_ERROR_REPORT_INFINITE_WAIT
+static void prv_create_infwait_err_rec(int pid, int state, unsigned long task_addr) {
+	infinty_waitdata_t param;
+	int ret;
+	param.pid = pid;
+	ret = ioctl(g_err_report_fd, ERIOC_GET_BACKTRACE, &param);
+	if (ret < 0) {
+		nwerr_vdbg("ioctl failed\n");
+		return;
+	}
+	error_infwait_data_t send_data;
+	struct timeval ts;
+	char sendbuf[ERR_BUFLEN];
+	int i;
+	gettimeofday(&ts, NULL);
+	send_data.error_type = ERRTYPE_HANGING;
+	send_data.ncalls = param.ncalls;
+	send_data.task_state = param.task_state;
+	send_data.task_addr = param.entry;
+	for (i = 0; i < CONFIG_ERROR_REPORT_BACKTRACE_MAX_DEPTH; i++) {
+		send_data.backtrace[i] = param.backtrace[i];
+	}
+	send_data.timestamp.tv_sec = ts.tv_sec;
+	send_data.timestamp.tv_usec = ts.tv_usec;
+	memcpy(sendbuf, (void *) &send_data, sizeof(error_infwait_data_t));
+	sendbuf[sizeof(error_infwait_data_t)] = 0;
+	/* Send error record to default endpoint */
+	error_report_send(0, sendbuf, sizeof(error_infwait_data_t) + 1);
+}
+#endif
+
+
+static int prv_check_infinite_wait(void *args) {
+	struct timespec ts;
+	struct timeval t1;
+	int ret;
+
+	while (1) {
+		ret = ioctl(g_err_report_fd, ERIOC_CHECK_INFWAIT, NULL);
+		if (ret < 0) {
+			nwerr_vdbg("ioctl failed\n");
+			goto done;
+		}
+		gettimeofday(&t1, NULL);
+		/* Logic to check every thread state whether it is in WAITSEM or WAITSIG mode */
+		nwerr_vdbg("%lu: sem_timedwait for %lu seconds\n", t1.tv_sec, ts.tv_sec);
+		ts.tv_sec = t1.tv_sec + CONFIG_ERROR_REPORT_INFINITE_CHECK_TIMER;
+		ts.tv_nsec = t1.tv_usec * 1000;
+		ret = sem_timedwait(&g_err_info.infinity_loop_exit, &ts);
+		gettimeofday(&t1, NULL);
+		nwerr_vdbg("%lu: sem_timedwait returned with value %d\n", t1.tv_sec, ret);
+		if (!ret) {
+			break;
+		} else if ((ret < 0) && (get_errno() != ETIMEDOUT)) {
+			printf("exiting with errno %d", get_errno());
+			goto done;
 		}
 	}
 	return 0;
+done:
+	sem_destroy(&g_err_info.infinity_loop_exit);
+	return ret;
 }
 
 err_status_t error_report_init(void)
 {
+	if (g_err_info.fsm == ERRSTATE_INITIALIZED) {
+		return ERR_SUCCESS;
+	}
 	pthread_mutex_init(&g_err_info.err_mutex, NULL);
 	g_err_info.fsm = ERRSTATE_INITIALIZED;
+#if CONFIG_ERROR_REPORT_INFINITE_WAIT
+	sem_init(&g_err_info.infinity_loop_exit, 0, 0);
+	error_report_drv_register(prv_create_infwait_err_rec);
```

Comment:
`error_report_drv_register` function is in kernel context. But external and apps folders are in user context. You should not call this from here. Please refer anothers.

Reply:
Duly noted, will implement `ioctl` command for setting the callback instead. Thank you.